okhttp/src/main/java/okhttp3/Authenticator.java

```diff
@@ -52,11 +52,33 @@
     @Override public Request authenticate(Route route, Response response) {
       return null;
     }
+
+    @Override
+    public Request authenticatePreemptive(Route route, Request request) throws IOException {
+      return null;
+    }
+
+    @Override
+    public boolean isPreemptive() {
+      return false;
+    }
   };
 
   /**
    * Returns a request that includes a credential to satisfy an authentication challenge in {@code
    * response}. Returns null if the challenge cannot be satisfied.
    */
   Request authenticate(Route route, Response response) throws IOException;
+
+  /**
+   * Return a request that is a copy of the original request plus authentication header.
```

Comment:
@yschimke All I need is basic proxy authentication. Maybe it would make more sense to add another interface ProxyAuthenticator that extends Authenticator. Anyway, my comment should be added to the implementation. I am thinking of adding BasicAuthenticator class.